Fix materialized CTE plan issue

kryonix opened the pull request:
This PR fixes an issue introduced in #10878, mentioned in https://github.com/duckdb/duckdb/pull/10878#issuecomment-2084676995. The previous PR changed the way materialized CTEs are placed in plans. However, the CTE map was represented by an `case_insensitive_map_t`, which internally uses `std::unordered_map`. This is problematic, because we have to keep the insertion order of CTEs.

This PR changes `case_insensitive_map_t` to `std::map`, which preserves the insertion order of CTE map entries.

---

An alternative design could be to create a CTE dependency graph to be used in `Transformer::TransformMaterializedCTE`. However, I don't think that this is necessary at this point.

kryonix committed:
Fix materialized CTE plan issue

Tmonster reviewed (approved):
Thank you for the quick fix! Would you mind adding the reproduction script as a micro regression test?
If CI passes looks good to me 👍

kryonix committed:
Add stacked materialized CTEs micro benchmark

kryonix committed:
Fix generated files

Mytherin reviewed (commented):
Thanks for the PR!

While `std::map` is ordered, it is ordered in alphabetical order, not in insertion order. This happens to be correct for the given benchmark because the CTE names are alphabetical (`mat_t1, mat_t2`, etc) - but does not hold in general. If we change the names in the query so that they are sorted in descending order, the performance regression happens again:

```sql
with
Hmat_t1 as materialized (select * from t1),
Gmat_t2 as materialized (select * from Hmat_t1),
Fmat_t3 as materialized (select * from Gmat_t2 where id not in (
                        select id % 20 from Gmat_t2
                        )),
Emat_t4 as materialized (select * from Hmat_t1 where id not in (
						select (id % 20) + 20 from Gmat_t2
						UNION ALL
						select (id % 20) + 40 from Fmat_t3
						)),
Dmat_t5 as materialized (select * from Hmat_t1 where id not in (
						select (id % 20) + 20 from Gmat_t2
						UNION ALL
						select (id % 20) + 40 from Fmat_t3
						UNION ALL
						select (id % 20) + 60 from Emat_t4
						)),
Cmat_t6 as materialized (select * from Hmat_t1 where id not in (
						select (id % 20) + 20 from Gmat_t2
						UNION ALL
						select (id % 20) + 40 from Fmat_t3
						UNION ALL
						select (id % 20) + 60 from Emat_t4
						UNION ALL
						select (id % 20) + 80 from Dmat_t5
						)),
Bmat_t7 as materialized (select * from Hmat_t1 where id not in (
						select (id % 20) + 20 from Gmat_t2
						UNION ALL
						select (id % 20) + 40 from Fmat_t3
						UNION ALL
						select (id % 20) + 60 from Emat_t4
						UNION ALL
						select (id % 20) + 80 from Dmat_t5
						UNION ALL
						select (id % 20) + 80 from Cmat_t6
						)),
Amat_t8 as materialized (select * from Hmat_t1 where id not in (
						select (id % 20) + 20 from Gmat_t2
						UNION ALL
						select (id % 20) + 40 from Fmat_t3
						UNION ALL
						select (id % 20) + 60 from Emat_t4
						UNION ALL
						select (id % 20) + 80 from Dmat_t5
						UNION ALL
						select (id % 20) + 80 from Cmat_t6
						UNION ALL
						select (id % 20) + 80 from Bmat_t7
						))
Select * from Hmat_t1 UNION ALL
select * from Gmat_t2 UNION ALL
select * from Fmat_t3 UNION ALL
select * from Emat_t4 UNION ALL
select * from Dmat_t5 UNION ALL
select * from Cmat_t6 UNION ALL
select * from Bmat_t7 UNION ALL
select * from Amat_t8;
```

In order to keep insertion order we would need to turn the `map` into a `vector<unique_ptr<CommonTableExpressionInfo>>`, with perhaps a `case_insensitive_map_t<string, idx_t>` to do quick name-based lookups.

kryonix commented:
Makes sense! I will change that accordingly.

kryonix committed:
Replace CTE map with vector for storage and map for lookups

kryonix committed:
Adapt sqlsmith to cte map changes

kryonix committed:
Run generate storage version

kryonix commented:
I've added a new field to the CTE map and to the serialization, too, as @Mytherin suggested. Hence I had to run the generate storage version script—which states that the storage version should be incremented. Is this necessary in this case?

Mytherin commented:
> I've added a new field to the CTE map and to the serialization, too, as @Mytherin suggested. Hence I had to run the generate storage version script—which states that the storage version should be incremented. Is this necessary in this case?

Hm, no actually we cannot do that anymore since that breaks backwards/forwards compatibility for views containing CTEs. Ideally we would keep the on-disk serialization the same. The previous serialization [serializes as a list of key/value pairs](https://github.com/duckdb/duckdb/blob/670cd341249e266de384e0341f200f4864b41b27/src/include/duckdb/common/serializer/serializer.hpp#L235). We can serialize the current representation in the same way - as long as we make sure to order the CTEs correctly (i.e. according to insertion order, not random/alphabetical order).

I think the cleanest way of doing this code-wise is to create a new class that represents the "insertion-order preserving map", e.g. `insertion_order_preserving_map<string, unique_ptr<CommonTableExpressionInfo>>`. Internally this can be represented using the vector + unordered_map combo. We can then special case the serialization/[deserialization](https://github.com/duckdb/duckdb/blob/670cd341249e266de384e0341f200f4864b41b27/src/include/duckdb/common/serializer/deserializer.hpp#L270) for this so that it is identical to the unordered_map/map.

Let me know if you want to pick this up - otherwise we can have a go at this as well.

kryonix committed:
Revert "Run generate storage version"

This reverts commit 5dcba3df87b915ea066ec9f706becba04affee6e.

kryonix committed:
Add InsertionOrderPreservingMap for CTEs

This commit adds an insertion order preserving map,
while keeping the serialization format of a regular map.

kryonix commented:
> Hm, no actually we cannot do that anymore since that breaks backwards/forwards compatibility for views containing CTEs.

Yeah, I figured that. I had something along the lines of "the storage format is fixed now" in the back of my mind ;-)

> I think the cleanest way of doing this code-wise is to create a new class that represents the "insertion-order preserving map"

I agree—And did just that. Thanks for the code pointers, those were really helpful.

kryonix committed:
Remove unused KEY_TYPE definition

Mytherin commented on src/include/duckdb/common/insertion_order_preserving_map.hpp:
```diff
@@ -0,0 +1,39 @@
+namespace duckdb {
+
+template <typename K, typename V>
+class InsertionOrderPreservingMap : public vector<V> {
```
Inheriting from `vector` seems like it is a bit error-prone here, since the `map_idx` and the vector could go out-of-sync quite easily (there is an invariant that `map_idx.size() == vector.size()`, but that is not enforced by the API here).

Perhaps we can instead have two private member variables that represent the map/vector, and implement the relevant methods that are necessary for the CTE map so that the invariant is maintained?

I think a nice interface here would be if the `InsertionOrderPreservingMap` behaves like a `std::map` in that we can access members using `operator[]`, or `.find()`, but when we iterate over the map (i.e. using `for(auto &entry : map)`) the entries are iterated over in insertion-order.

`K` should also always be an `idx_t`, so we should not need to parameterize that, right? Currently the actual key is always a string (as part of `case_insensitive_map_t`).

kryonix replied:
Great idea! You are right, the map and vector should never get out of sync. An explicit interface should be the way to go here.

Mytherin reviewed (commented):
Thanks! This looks great. One more comment:

kryonix committed:
Add proper interface to insertion order preserving map

This commit simplifies usage of the insertion order preserving
map massively. Instead of exposing the vector and map directly,
only the necessary functions are exposed through an appropriate
interface. This also hinders users to accidentally corrupt the map.

kryonix committed:
Fix implicit conversion signedness change issue

kryonix committed:
Fix tidy-check

kryonix committed:
Fix format check

kryonix commented:
I think this PR is ready for another round of reviewing. I've changed the insertion order map as proposed and provided an stl-style interface for it. This ensures that the vector and map stay synchronized, simplifies usage, and automatically does the _right thing_ when used in `for-each` loops.

Mytherin committed:
Merge branch 'main' into cte_regression

Mytherin commented:
Thanks! LGTM - just had to resolve some merge conflicts

Mytherin committed:
Fix merge conflicts

Mytherin committed:
Fix merge conflicts
